x2gbfs/providers/noi.py

```diff
@@ -0,0 +1,124 @@
+import logging
+from typing import Dict, Optional, Tuple
+
+from x2gbfs.gbfs.base_provider import BaseProvider
+import requests, re
+
+logger = logging.getLogger(__name__)
+
+class Counter:
+    """
+    A utility class to group certain cars by their station id.
+    """
+    cars = {}
+    def __init__(self):
+        pass
+
+    def count_car(self, station_id, type):
+        if self.cars.get(station_id) is None:
+            self.cars[station_id] = { type: 1 }
+        elif self.cars[station_id].get(type) is None:
+            self.cars[station_id][type] = 1
+        else:
+            count = self.cars[station_id][type]
+            self.cars[station_id][type] = count + 1
+
+    def total_cars(self, station_id):
+        cars = self.cars.get(station_id)
+        if cars is None:
+            return 0
+        else:
+            count = 0
+            for i in cars:
+                count = count + cars[i]
+        return count
+
+class NoiProvider(BaseProvider):
+    STATION_URL = "https://mobility.api.opendatahub.com/v2/flat%2Cnode/CarsharingStation?limit=500&offset=0&shownull=false&distinct=true"
+    CAR_URL = "https://mobility.api.opendatahub.com/v2/flat%2Cnode/CarsharingCar?limit=500&offset=0&shownull=false&distinct=true"
+
+    def __init__(self):
+        pass
+
+    def load_vehicles(self, default_last_reported: int) -> Tuple[Optional[Dict], Optional[Dict]]:
+        response = requests.get(self.CAR_URL, timeout=20)
+        raw_cars = response.json()["data"]
+        types = {}
+        for i in raw_cars:
+
+                id = self.extract_type_id(i)
+                types[id] = {
+                    # See https://github.com/MobilityData/gbfs/blob/v2.3/gbfs.md#vehicle_typesjson
+                    'vehicle_type_id': id,
+                    'form_factor': 'car',
```

Comment:
You could set station_id to pcode here. In addition to providing more information, you then would not even need to count the `num_bikes_available` and `vehicle_types_available`, which can deduced by the `GbfsTransformer`

Reply:
I don't think I understand: I need to set the station id on the vehicle type?

Reply:
Ah sorry, no, you would need to create vehicles in addition.

Reply:
I assumed that the vehicles that I return here are free floating ones. But you're saying that if i set a station_id they will be assigned to a station?

Reply:
Yes, GBFS supports station based vehicles also. For these, no coordinates need to be provided as they are implicitly known via the station. But you'd need to check, if the API provides the required information (or assume some sensible defaults). Checking the generated results via https://gbfs-validator.mobilitydata.org also is a good idea.

Reply:
Ah, so it's GBFS feature, not a x2gbfs feature?

I do check the feed with the validator as part of my development cycle and it's valid: https://gbfs-validator.netlify.app/validator?url=https%3A%2F%2Fleonard.io%2Fgbfs%2Fnoi%2Fgbfs.json

Reply:
x2gbfs deduces `num_bikes_available` and `vehicle_types_available` and renders it in gbfs, for which it is required. x2gbfs respects `is_reserved` state and might list (only currently) unavailable vehicle_types at a station with a count of zero.

Reply:
So I need to return the vehicles together with the vehicle types? I assumed that they then are published in free_bikes_status. Is this taken care of?

Reply:
That depends on the vehicle_types `return_constraint`. For carsharing, usually it's usually `roundtrip_station`, for station based bike sharing it's often `any_station`, see also [GBFSv2.3 spec](https://github.com/MobilityData/gbfs/blob/v2.3/gbfs.md#vehicle_typesjson)